File: Solutions/9739/python/main.py
def calculate_largest_rectangle_area(heights):
    stack = []
    max_area = 0
    i = 0

    while i < len(heights):
        if not stack or heights[stack[-1]] <= heights[i]:
            stack.append(i)
            i += 1
        else:
            top = stack.pop()
            area_with_top = heights[top] * (i if not stack else i - stack[-1] - 1)

            if max_area < area_with_top:
                max_area = area_with_top

    while stack:
        top = stack.pop()
        area_with_top = heights[top] * (i if not stack else i - stack[-1] - 1)

        if max_area < area_with_top:
            max_area = area_with_top

    return max_area


if __name__ == "__main__":
    n = int(input())
    heights = list(map(int, input().split()))

    result = calculate_largest_rectangle_area(heights)
    print(result)

File: Solutions/9739/c++/main.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

int calculateLargestRectangleArea(const vector<int>& heights) {
    stack<int> s;
    int maxArea = 0;
    int i = 0;

    while (i < heights.size()) {
        if (s.empty() || heights[s.top()] <= heights[i]) {
            s.push(i++);
        } else {
            int top = s.top();
            s.pop();
            int areaWithTop = heights[top] * (s.empty() ? i : i - s.top() - 1);

            if (maxArea < areaWithTop) {
                maxArea = areaWithTop;
            }
        }
    }

    while (!s.empty()) {
        int top = s.top();
        s.pop();
        int areaWithTop = heights[top] * (s.empty() ? i : i - s.top() - 1);

        if (maxArea < areaWithTop) {
            maxArea = areaWithTop;
        }
    }

    return maxArea;
}

int main() {
    int n;
    cin >> n;

    vector<int> heights(n);

    for (int i = 0; i < n; i++) {
        cin >> heights[i];
    }

    int result = calculateLargestRectangleArea(heights);
    cout << result << endl;

    return 0;
}
